README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# discursive

<!-- badges: start -->
<!-- badges: end -->

This package implements a simple but powerful framework to measure
political sophistication based on open-ended survey responses.
Discursive sophistication captures the complexity of individual attitude
expression by quantifying its relative size, range, and constraint. For
more information on the measurement approach see: Kraft, Patrick W.
2023. “Women Also Know Stuff: Challenging the Gender Gap in Political
Sophistication.” American Political Science Review (forthcoming).

## Installation

`discursive` is now available on CRAN! You can install the latest
version with:

``` r
install.packages("discursive")
```

Alternatively, you can install the development version of discursive
from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("pwkraft/discursive")
```

## Example

``` r
library(discursive)
cces$discursive <- discursive(data = cces,
                              openends = c(paste0("oe0", 1:9), "oe10"),
                              meta = c("age", "educ_cont", "pid_cont", "educ_pid", "female"),
                              args_prepDocuments = list(lower.thresh = 10, verbose = FALSE),
                              args_stm = list(K = 25, seed = 12345, verbose = FALSE),
                              dictionary = dict_sample, progress = FALSE)$output$discursive
#> Building corpus... 
#> Converting to Lower Case... 
#> Removing punctuation... 
#> Removing stopwords... 
#> Removing numbers... 
#> Stemming... 
#> Creating Output...
head(cces)
#>   age female educ_cont pid_cont educ_pid
#> 1  58      1         2        1        2
#> 2  70      1         4        1        4
#> 3  23      1         4        6       24
#> 4  29      1         6        2       12
#> 5  28      1         4        2        8
#> 6  57      1         2        2        4
#>                                                                                                                                                                                                                                   oe01
#> 1                                                                                                                                                                                          Check for past arrests or mental conditions
#> 2 Necessary to provide adequate information for anyone who wishes to purchase a weapon. This will prevent anyone not of sound mind to have a weapon. Also, anyone who has committed a serious crime should need further investigation.
#> 3                                                                                                                                                                                      so guns do not get into the wrong peoples hands
#> 4                                                                                                                                                                                                         Less illegal guns on streets
#> 5                                                                                                                                                                            We want to protect the greater society by double checking
#> 6                                                                                                                                                to make sure the person doesn't have a criminal history or they are mentally unstable
#>                                                                                                oe02
#> 1                                                                            More and quicker sales
#> 2    These thinkers fear that anything at all will weaken the second amendment to the Constitution.
#> 3                                                                                to use for hunting
#> 4                                                Unfair everyone should have the right to beat arms
#> 5 That background checks shouldn’t matter because they are just wanting a gun for hunting purposes.
#> 6                      it is against everyone's right to bear arms and it is an invasion of privacy
#>                                                                                                                         oe03
#> 1                                                                                                         Abortion is murder
#> 2 Mostly from religious organizations some of whom refer to the bible. They may not understand that it's not about religion.
#> 3                                                                women have the right to make decisions about their own body
#> 4                                                                                     People will be more careful having sex
#> 5                                                                                                    That it is a human life
#> 6                                                                                                 it is murder of the unborn
#>                                                                                                               oe04
#> 1                                                                                                  Right to decide
#> 2 A woman should have the absolute right to decide what happens to her body as do men. Roe v Wade provides choice.
#> 3                                                                            you should not be able to kill a baby
#> 4                                                                                               Demeaning to women
#> 5                                                       It is the women’s right to do what she wants with her body
#> 6                                                      women have the right to decide what to do with their bodies
#>                                                                                                                                                                                 oe05
#> 1                                                                                                                                                     Provide a better life for them
#> 2 This is the only fair and just way to provide legal status to these people who know nothing but the US, have been educated here and work to provide a better life for them and us.
#> 3                                                                                                                                                    America is the home for chances
#> 4                                                                                                                                        People are people let them pursue happiness
#> 5                                                                                                                                               This is their home and all they know
#> 6                                                                                                  American is built on the idea of being a melting pot with open arms to immigrants
#>                                                              oe06
#> 1                                    Overpopulating and less jobs
#> 2                                                                
#> 3                  people need to earn their way into our country
#> 4                                                We can help them
#> 5                                       They are still immigrants
#> 6 they drain the system resources and some of them are terrorists
#>                                                                                      oe07
#> 1                                                                       Is not affordable
#> 2                                               Costs too much and was a Democratic idea.
#> 3                                      everyone had a chance to be covered by health care
#> 4                                                                                        
#> 5                              It is costing us too much money and we have better options
#> 6 it forces people to purchase something against their will that they may or may not want
#>                                                                                    oe08
#> 1                                                    More money for insurance companies
#> 2 All Americans deserve healthcare just like most of the world's progressive countries.
#> 3                              not everyone thinks its a requirement to have healthcare
#> 4                                                     People need affordable healthcare
#> 5                                    Keep protecting the ones we are already protecting
#> 6                                                   it helps more people have insurance
#>                                                 oe09
#> 1                                                   
#> 2           Just another ridiculous Republican idea.
#> 3       We keep a line of trade with other countries
#> 4                                                   
#> 5 Since they are out of the country we should charge
#> 6 we have to pay them tariffs so they can pay us too
#>                                                                           oe10
#> 1                                                                             
#> 2 This just irritates our allies and hurts our workers, companies and farmers.
#> 3                              We close the line of trade with other countries
#> 4                                                          Less big government
#> 5                                                           It is just trading
#> 6                                    it isn't good for international relations
#>    discursive
#> 1  0.07362755
#> 2  1.27235839
#> 3  0.32848993
#> 4 -0.17371927
#> 5  0.79568885
#> 6  0.50372490
```

<!-- What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so: -->
<!-- ```{r cars} -->
<!-- summary(cars) -->
<!-- ``` -->
<!-- You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this. You could also use GitHub Actions to re-render `README.Rmd` every time you push. An example workflow can be found here: <https://github.com/r-lib/actions/tree/v1/examples>. -->
<!-- You can also embed plots, for example: -->
<!-- ```{r pressure, echo = FALSE} -->
<!-- plot(pressure) -->
<!-- ``` -->
<!-- In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN. -->


NEWS.md
# discursive (development version)

# discursive 0.1.1

* Initial CRAN release.
* Added a `NEWS.md` file to track changes to the package.


R/discursive_size.R
#' Compute the size component of discursive sophistication
#'
#' This function takes a data frame (`data`) containing a set of open-ended responses (`openends`) and additional arguments passed to [stm::textProcessor()] and [stm::prepDocuments()] to estimate a structural topic model via [stm::stm()]. The results of the the structural topic model are used to compute the relative number of topics raised in each open-ended response. The function returns a numeric vector of topic counts re-scaled to range from 0 to 1. See Kraft (2023) for details.
#'
#' @param data A data frame.
#' @param openends A character vector containing variable names of open-ended responses in `data`.
#' @param meta A character vector containing topic prevalence covariates included in `data`. See [stm::stm()] for details.
#' @param args_textProcessor A named list containing additional arguments passed to [stm::textProcessor()].
#' @param args_prepDocuments A named list containing additional arguments passed to [stm::prepDocuments()].
#' @param args_stm A named list containing additional arguments passed to [stm::stm()].
#' @param keep_stm Logical. If TRUE function returns output of [stm::textProcessor()], [stm::prepDocuments()], and [stm::stm()].
#' @param progress Logical. Shows progress bar if TRUE.
#'
#' @return A list containing the size component of discursive sophistication as well as the output of [stm::textProcessor()], [stm::prepDocuments()], and [stm::stm()].
#' @export
#' @import stm
#' @import tm
#' @import SnowballC
#'
#' @examples
#' \donttest{discursive_size(data = cces,
#'                 openends = c(paste0("oe0", 1:9), "oe10"),
#'                 meta = c("age", "educ_cont", "pid_cont", "educ_pid", "female"),
#'                 args_prepDocuments = list(lower.thresh = 10),
#'                 args_stm = list(K = 25, seed = 12345))}
discursive_size <- function(data, openends, meta,
                            args_textProcessor = NULL,
                            args_prepDocuments = NULL,
                            args_stm = NULL,
                            keep_stm = TRUE,
                            progress = TRUE){

  ## Check input
  if(!is.data.frame(data)) stop("data must be a data frame.")
  if(!is.character(openends)) stop("openends must be a character vector.")
  if(!is.character(meta)) stop("meta must be a character vector.")

  ## Remove observations with missing values on metadata
  nomis_id <- which(!apply(data[, meta], 1, anyNA))
  nomis <- data[nomis_id, ]

  ## Combine open-ended items
  resp <- apply(nomis[, openends], 1, paste, collapse = " ")

  ## Prepare for stm: textProcessor
  if(is.null(args_textProcessor)) {
    args_textProcessor <- list(documents = resp,
                               metadata = nomis[, meta])
  } else {
    args_textProcessor <- as.list(args_textProcessor)
    if("documents" %in% names(args_textProcessor)) warning("documents in args_textProcessor was replaced internally.")
    if("metadata" %in% names(args_textProcessor)) warning("metadata in args_textProcessor was replaced internally.")
    args_textProcessor$documents <- resp
    args_textProcessor$metadata <- nomis[, meta]
  }
  processed <- do.call(stm::textProcessor, args_textProcessor)

  ## Prepare for stm: prepDocuments
  if(is.null(args_prepDocuments)) {
    args_prepDocuments <- list(documents = processed$documents,
                               vocab = processed$vocab,
                               meta = as.matrix(processed$meta))
  } else {
    args_prepDocuments <- as.list(args_prepDocuments)
    if("documents" %in% names(args_prepDocuments)) warning("documents in args_prepDocuments was replaced internally.")
    if("vocab" %in% names(args_prepDocuments)) warning("vocab in args_prepDocuments was replaced internally.")
    if("meta" %in% names(args_prepDocuments)) warning("meta in args_prepDocuments was replaced internally.")
    args_prepDocuments$documents <- processed$documents
    args_prepDocuments$vocab <- processed$vocab
    args_prepDocuments$meta <- as.matrix(processed$meta)
  }
  out <- do.call(stm::prepDocuments, args_prepDocuments)

  ## Remove discarded observations from nomis_id
  if(length(processed$docs.removed)>0) nomis_id <- nomis_id[-processed$docs.removed]
  if(!is.null(out$docs.removed)) nomis_id <- nomis_id[-out$docs.removed]

  ### Fit stm
  if(is.null(args_stm)) {
    args_stm <- list(documents = out$documents,
                     vocab = out$vocab,
                     prevalence = out$meta)
  } else {
    args_stm <- as.list(args_stm)
    if("documents" %in% names(args_stm)) warning("documents in args_stm was replaced internally.")
    if("vocab" %in% names(args_stm)) warning("vocab in args_stm was replaced internally.")
    if("prevalence" %in% names(args_stm)) warning("prevalence in args_stm was replaced internally.")
    args_stm$documents <- out$documents
    args_stm$vocab <- out$vocab
    args_stm$prevalence <- out$meta
  }
  stm_fit <- do.call(stm::stm, args_stm)

  ### Compute number of considerations
  size <- ntopics(stm_fit, out, progress)

  ### Add NAs for missing observations
  size_na <- rep(NA, nrow(data))
  size_na[nomis_id] <- size

  ## Return output
  if(!keep_stm){
    processed <- NULL
    out <- NULL
    stm_fit <- NULL
  }
  list(size = size_na,
       textProcessor = processed,
       prepDocuments = out,
       stm = stm_fit)
}


#' Compute number of topics based on stm results
#'
#' This function takes a structural topic model output estimated via [stm::stm()] as well as the underlying set of documents created via [stm::prepDocuments()] to compute the relative number of topics raised in each open-ended response. The function returns a numeric vector of topic counts re-scaled to range from 0 to 1. See Kraft (2023) for details.
#'
#' @param x A structural topic model estimated via [stm::stm()].
#' @param docs A set of documents used for the structural topic model; created via [stm::prepDocuments()].
#' @param progress Logical. Shows progress bar if TRUE.
#'
#' @return A numeric vector with the same length as the number of documents in `x` and `docs`.
#' @import stm
#' @import utils
#' @export
#'
#' @examples
#' \donttest{meta <- c("age", "educ_cont", "pid_cont", "educ_pid", "female")
#' openends <- c(paste0("oe0", 1:9), "oe10")
#' cces$resp <- apply(cces[, openends], 1, paste, collapse = " ")
#' cces <- cces[!apply(cces[, meta], 1, anyNA), ]
#' processed <- stm::textProcessor(cces$resp, metadata = cces[, meta])
#' out <- stm::prepDocuments(processed$documents, processed$vocab, processed$meta, lower.thresh = 10)
#' stm_fit <- stm::stm(out$documents, out$vocab, prevalence = as.matrix(out$meta), K=25, seed=12345)
#' ntopics(stm_fit, out)}
ntopics <- function(x, docs, progress = TRUE){
  if(!inherits(x, "STM")) stop("x must be an STM object")

  ## P(t|X): probability of topic t given covariates X [nobs,ntopics]
  pt_x <- x$theta

  ## P(w|t): probability of word w given topic t [ntopics,nwords]
  pw_t <- exp(x$beta$logbeta[[1]])
  # logbeta: list containing the log of the word probabilities for each topic

  ## P(w|X) = sum_j(P(w|t_j)P(t_j|X)): probability of word w across all topics [nobs,nwords]
  pw_x <- pt_x %*% pw_t

  ## P(t_i|w,X) = P(w|t_i)P(t_i|X)/P(w|X): probability of topic t given word w and covariates X
  nobs <- nrow(pt_x)
  nwords <- ncol(pw_t)
  ntopics <- ncol(pt_x)
  pt_wx <- array(NA, c(nobs, nwords, ntopics))

  if(progress) cat("\nComputing P(t_i|w,X):\n")
  if(progress) pb <- utils::txtProgressBar(min = 1, max = nwords, style = 3)
  for(w in 1:nwords){
    for(t in 1:ntopics){
      pt_wx[,w,t] <- pw_t[t,w] * pt_x[,t] / pw_x[,w]
    }
    if(progress) utils::setTxtProgressBar(pb, w)
  }
  if(progress) close(pb)

  ## compute sophistication components
  if(progress) cat("\nComputing relative topic counts:\n")
  ntopics <- rep(NA, nobs)
  if(progress) pb <- utils::txtProgressBar(min = 1, max = nobs, style = 3)
  for(n in 1:nobs){
    maxtopic_wx <- apply(pt_wx[n,,],1,which.max) # which topic has the highest probability for each word (given X)
    ntopics[n] <- length(unique(maxtopic_wx[docs$documents[[n]][1,]])) # number of topics in response
    if(progress) utils::setTxtProgressBar(pb, n)
  }
  if(progress) close(pb)
  if(progress) cat("\n\n")

  ## return rescaled measure
  ntopics/max(ntopics)
}
